### PR TITLE
feat: emode category for v3 config engine

### PR DESCRIPTION
**Pull request by brotherlymite**
- splits the config engine into different engine libraries
- added emode category update

### CODE REVIEW

File: src/v3-config-engine/AaveV3ConfigEngine.sol [hunk 1]
```diff
@@ -3,10 +3,14 @@ pragma solidity ^0.8.12;
 
 import {ConfiguratorInputTypes, DataTypes} from 'aave-address-book/AaveV3.sol';
```
Comment (eboadom) [inline src/v3-config-engine/AaveV3ConfigEngine.sol]: Unused imports, better review all

File: src/v3-config-engine/AaveV3ConfigEngine.sol [hunk 2]
```diff
@@ -15,98 +19,106 @@ import './IAaveV3ConfigEngine.sol';
  * - It is planned to be used via delegatecall, by any contract having appropriate permissions to
```
Comment (eboadom) [inline src/v3-config-engine/AaveV3ConfigEngine.sol:18]: Add comments that it is completely mandatory for this contract to be stateless

File: src/v3-config-engine/AaveV3ConfigEngine.sol [hunk 3]
```diff
@@ -15,98 +19,106 @@ import './IAaveV3ConfigEngine.sol';
...
-    address[] ids;
-    Basic[] basics;
-    Borrow[] borrows;
-    Collateral[] collaterals;
-    Caps[] caps;
-    IV3RateStrategyFactory.RateStrategyParams[] rates;
-  }
+  using Address for address;
 
   struct Basic {
```
Comment (eboadom) [inline src/v3-config-engine/AaveV3ConfigEngine.sol]: Unused

File: src/v3-config-engine/AaveV3ConfigEngine.sol [hunk 4]
```diff
@@ -126,529 +138,118 @@ contract AaveV3ConfigEngine is IAaveV3ConfigEngine {
...
-      AssetsConfig({
-        ids: ids,
-        caps: new Caps[](0),
-        basics: new Basic[](0),
-        borrows: new Borrow[](0),
-        collaterals: collaterals,
-        rates: new IV3RateStrategyFactory.RateStrategyParams[](0)
-      });
+  /// @inheritdoc IAaveV3ConfigEngine
+  function updateEModeAssets(EModeAssetUpdate[] calldata updates) external {
```
Comment (eboadom) [inline src/v3-config-engine/AaveV3ConfigEngine.sol]: I would name it better as `updateAssetsEMode`

**Comment by eboadom** [comment]
https://github.com/bgd-labs/aave-helpers/pull/97/files#diff-05162a646e275af7d24e687b3b7b8850f7957757e2bbfdfda6e3b28f641eed78R30

It is pretty legacy keeping using the `LISTING_ENGINE` instead of `CONFIG_ENGINE`, and now with the introduction of a ListingEngine library, really confusing.
I would say better to replace, checking the dependencies we have of using the `LISTING_ENGINE` getter

File: src/v3-config-engine/AaveV3PayloadBase.sol [hunk 5]
```diff
@@ -14,10 +14,14 @@ import {EngineFlags} from './EngineFlags.sol';
...
  *   - Updates of caps (supply cap, borrow cap).
  *   - Updates of price feeds
+ *   - Updates of interest rate strategies
  *   - Updates of borrow parameters (flashloanable, stableRateModeEnabled, borrowableInIsolation, withSiloedBorrowing, reserveFactor)
  *   - Updates of collateral parameters (ltv, liq threshold, liq bonus, liq protocol fee, debt ceiling)
+ *   - Updates of emode category parameters (ltv, liq threshold, liq bonus, price source, label)
+ *   - Updates of emode category of assets (e-mode id)
  * @author BGD Labs
  */
 abstract contract AaveV3PayloadBase {
```
Comment (eboadom) [inline src/v3-config-engine/AaveV3PayloadBase.sol:28]: Let's add also comment that it is mandatory for this contract to be stateless

File: src/v3-config-engine/IAaveV3ConfigEngine.sol [hunk 6]
```diff
@@ -155,6 +153,38 @@ interface IAaveV3ConfigEngine {
     uint256 reserveFactor;
   }
 
+  /**
+   * @dev Example (mock):
+   * EModeAssetUpdate({
+   *   asset: AaveV3EthereumAssets.AAVE_UNDERLYING,
```
Comment (eboadom) [inline src/v3-config-engine/IAaveV3ConfigEngine.sol]: Let's put reasonable example, so some ETH LST, not AAVE

File: src/v3-config-engine/IAaveV3ConfigEngine.sol [hunk 7]
```diff
@@ -155,6 +153,38 @@ interface IAaveV3ConfigEngine {
...
+   */
+  struct EModeAssetUpdate {
+    address asset;
+    uint8 eModeCategory;
+  }
+
+  /**
+   * @dev Example (mock):
+   * EModeCategoryUpdate({
+   *   eModeCategory: 1, // STABLECOINS
```
Comment (eboadom) [inline src/v3-config-engine/IAaveV3ConfigEngine.sol]: On previous examples `1` is used as ETH correlated, so better to keep consistency changing here, or everywhere else

File: src/v3-config-engine/AaveV3ConfigEngine.sol [hunk 8]
```diff
@@ -15,98 +19,106 @@ import './IAaveV3ConfigEngine.sol';
...
-  }
-
-  struct Collateral {
-    uint256 ltv; // Only considered if liqThreshold > 0. With 2 digits precision, `10_00` for 10%. Should be lower than liquidationThreshold
-    uint256 liqThreshold; // If `0`, the asset will not be enabled as collateral. Same format as ltv, and should be higher
-    uint256 liqBonus; // Only considered if liqThreshold > 0. Same format as ltv
-    uint256 debtCeiling; // Only considered if liqThreshold > 0. In USD and without decimals, so 100_000 for 100k USD debt ceiling
-    uint256 liqProtocolFee; // Only considered if liqThreshold > 0. Same format as ltv
-    uint256 eModeCategory;
+  struct EngineLibraries {
```
Comment (eboadom) [inline src/v3-config-engine/AaveV3ConfigEngine.sol]: Better to declare this and `EngineConstants` on the interface, that way on other libs you don't need to import the config engine, just the interface

**Review by rustboyar — APPROVED** [review]

**Pull request merged**